File: modlab/src/ui/PackCard.cpp
#include "PackCard.h"
#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QPainter>
#include <QMouseEvent>
#include <QGraphicsDropShadowEffect>
#include <QStyle>

PackCard::PackCard(const Pack& pack, QWidget* parent)
    : QFrame(parent)
{
    setObjectName("PackCard");
    setFixedWidth(240);
    setCursor(Qt::PointingHandCursor);
    buildLayout(pack);
}

void PackCard::buildLayout(const Pack& pack) {
    m_key = pack.key;

    auto* root = new QVBoxLayout(this);
    root->setContentsMargins(15, 14, 15, 14);
    root->setSpacing(0);

    // ---- Header row: ident tile + status pill ----
    auto* headRow = new QHBoxLayout;
    headRow->setContentsMargins(0,0,0,10);

    m_identLabel = new QLabel(pack.abbr(), this);
    m_identLabel->setFixedSize(40, 40);
    m_identLabel->setAlignment(Qt::AlignCenter);
    m_identLabel->setObjectName("IdentTile");
    // Coloring is done in paintEvent / updatePack

    m_statusLabel = new QLabel(statusLabel(pack.status), this);
    m_statusLabel->setAlignment(Qt::AlignRight | Qt::AlignVCenter);

    headRow->addWidget(m_identLabel);
    headRow->addStretch();
    headRow->addWidget(m_statusLabel);
    root->addLayout(headRow);

    // ---- Pack name ----
    m_nameLabel = new QLabel(pack.name, this);
    m_nameLabel->setObjectName("PackName");
    root->addWidget(m_nameLabel);
    root->addSpacing(4);

    // ---- Tag row ----
    m_tagRow = new QWidget(this);
    auto* tagLayout = new QHBoxLayout(m_tagRow);
    tagLayout->setContentsMargins(0,0,0,0);
    tagLayout->setSpacing(5);

    auto makeTag = [&](const QString& text) {
        QLabel* tag = new QLabel(text, m_tagRow);
        tag->setObjectName("TagPill");
        tagLayout->addWidget(tag);
    };
    makeTag(loaderName(pack.loader));
    makeTag(pack.mcVersion);
    if (pack.modCount > 0)
        makeTag(QString("%1 mods").arg(pack.modCount));
    tagLayout->addStretch();

    root->addWidget(m_tagRow);
    root->addSpacing(12);

    // ---- Separator line ----
    QFrame* sep = new QFrame(this);
    sep->setFrameShape(QFrame::HLine);
    sep->setStyleSheet("color: rgba(56,217,245,0.12);");
    sep->setFixedHeight(1);
    root->addWidget(sep);
    root->addSpacing(11);

    // ---- Action buttons ----
    auto* btnRow = new QHBoxLayout;
    btnRow->setSpacing(6);

    m_btnRun     = new QPushButton("Run",     this);
    m_btnInstall = new QPushButton("Install", this);
    m_btnDelete  = new QPushButton("Delete",  this);

    m_btnRun->setObjectName("BtnRun");
    m_btnInstall->setObjectName("BtnSecondary");
    m_btnDelete->setObjectName("BtnDanger");

    btnRow->addWidget(m_btnRun);
    btnRow->addWidget(m_btnInstall);
    btnRow->addWidget(m_btnDelete);
    root->addLayout(btnRow);

    // ---- Progress (hidden by default) ----
    m_progress = new QProgressBar(this);
    m_progress->setRange(0, 100);
    m_progress->setValue(pack.installProgress);
    m_progress->setTextVisible(false);
    m_progress->setFixedHeight(3);

    m_progressLbl = new QLabel(this);
    m_progressLbl->setObjectName("MutedText");
    m_progressLbl->setStyleSheet("font-size: 9px;");

    root->addSpacing(9);
    root->addWidget(m_progress);
    root->addSpacing(3);
    root->addWidget(m_progressLbl);

    applyStatus(pack.status);

    connect(m_btnRun,     &QPushButton::clicked, this, [this]{ emit runRequested(m_key); });
    connect(m_btnInstall, &QPushButton::clicked, this, [this]{ emit installRequested(m_key); });
    connect(m_btnDelete,  &QPushButton::clicked, this, [this]{ emit deleteRequested(m_key); });
}

void PackCard::applyStatus(PackStatus status) {
    bool installing = (status == PackStatus::Installing);
    m_progress->setVisible(installing);
    m_progressLbl->setVisible(installing);
    m_btnRun->setEnabled(status == PackStatus::Ready || status == PackStatus::NeedsUpdate);

    switch (status) {
        case PackStatus::Ready:
            m_statusLabel->setObjectName("StatusReady");
            m_btnInstall->setText("Install");
            break;
        case PackStatus::Installing:
            m_statusLabel->setObjectName("StatusBusy");
            m_btnInstall->setText("Reinstall");
            break;
        case PackStatus::NeedsUpdate:
            m_statusLabel->setObjectName("StatusWarn");
            m_btnInstall->setText("Update");
            break;
        case PackStatus::NotInstalled:
            m_statusLabel->setObjectName("StatusWarn");
            m_btnInstall->setText("Install");
            break;
    }
    // Force QSS re-evaluation after objectName change
    m_statusLabel->style()->unpolish(m_statusLabel);
    m_statusLabel->style()->polish(m_statusLabel);
}

void PackCard::updatePack(const Pack& pack) {
    m_nameLabel->setText(pack.name);
    m_identLabel->setText(pack.abbr());
    m_statusLabel->setText(statusLabel(pack.status));
    m_progress->setValue(pack.installProgress);
    m_progressLbl->setText(QString("Downloading... %1%").arg(pack.installProgress));
    applyStatus(pack.status);
    update();
}

void PackCard::setSelected(bool selected) {
    setProperty("selected", selected);
    style()->unpolish(this);
    style()->polish(this);
    update();
}

void PackCard::mousePressEvent(QMouseEvent* event) {
    QFrame::mousePressEvent(event);
    emit cardClicked(m_key);
}

void PackCard::paintEvent(QPaintEvent* event) {
    // Let QSS handle the card background; just add the top-shine strip
    QFrame::paintEvent(event);

    QPainter p(this);
    p.setRenderHint(QPainter::Antialiasing);

    // Top shine gradient strip
    QLinearGradient shine(0, 0, width(), 0);
    shine.setColorAt(0.0, Qt::transparent);
    shine.setColorAt(0.5, QColor(56, 217, 245, 30));
    shine.setColorAt(1.0, Qt::transparent);
    p.fillRect(0, 0, width(), 1, shine);

    // Inner top-glow gradient
    QLinearGradient topGlow(0, 0, 0, 40);
    topGlow.setColorAt(0.0, QColor(56, 217, 245, 10));
    topGlow.setColorAt(1.0, Qt::transparent);
    p.fillRect(0, 0, width(), 40, topGlow);
}

File: modlab-pyshit-client/ui/pack_card.py
from PySide6.QtCore import Qt, Signal
from PySide6.QtGui import QColor, QLinearGradient, QMouseEvent, QPainter
from PySide6.QtWidgets import QFrame, QHBoxLayout, QLabel, QProgressBar, QPushButton, QVBoxLayout, QWidget

from ..core.pack import Pack, PackStatus, loader_name, status_label


class PackCard(QFrame):
    run_requested = Signal(str)
    install_requested = Signal(str)
    delete_requested = Signal(str)
    card_clicked = Signal(str)

    def __init__(self, pack: Pack, parent=None):
        super().__init__(parent)
        self.setObjectName("PackCard")
        self.setFixedWidth(240)
        self.setCursor(Qt.PointingHandCursor)
        self.key = pack.key
        self._build_layout(pack)

    def _build_layout(self, pack: Pack) -> None:
        root = QVBoxLayout(self)
        root.setContentsMargins(15, 14, 15, 14)
        root.setSpacing(0)

        head_row = QHBoxLayout()
        head_row.setContentsMargins(0, 0, 0, 10)
        self.ident_label = QLabel(pack.abbr(), self)
        self.ident_label.setFixedSize(40, 40)
        self.ident_label.setAlignment(Qt.AlignCenter)
        self.ident_label.setStyleSheet(
            "font-family:'Exo 2'; font-weight:900; font-size:12px;"
            "background: qlineargradient(x1:0,y1:0,x2:0.5,y2:1,"
            "stop:0 rgba(0,200,180,0.25), stop:1 rgba(0,130,120,0.3));"
            "border-radius:8px; color:#00e5cc;"
        )
        self.status_label = QLabel(status_label(pack.status), self)
        self.status_label.setAlignment(Qt.AlignRight | Qt.AlignVCenter)
        head_row.addWidget(self.ident_label)
        head_row.addStretch()
        head_row.addWidget(self.status_label)
        root.addLayout(head_row)

        self.name_label = QLabel(pack.name, self)
        self.name_label.setObjectName("PackName")
        root.addWidget(self.name_label)
        root.addSpacing(4)

        self.tag_row = QWidget(self)
        tag_layout = QHBoxLayout(self.tag_row)
        tag_layout.setContentsMargins(0, 0, 0, 0)
        tag_layout.setSpacing(5)
        for text in [loader_name(pack.loader), pack.mc_version]:
            tag_layout.addWidget(self._make_tag(text))
        if pack.mod_count > 0:
            tag_layout.addWidget(self._make_tag(f"{pack.mod_count} mods"))
        tag_layout.addStretch()
        root.addWidget(self.tag_row)
        root.addSpacing(12)

        sep = QFrame(self)
        sep.setFrameShape(QFrame.HLine)
        sep.setStyleSheet("color: rgba(56,217,245,0.12);")
        sep.setFixedHeight(1)
        root.addWidget(sep)
        root.addSpacing(11)

        button_row = QHBoxLayout()
        button_row.setSpacing(6)
        self.run_button = QPushButton("Run", self)
        self.install_button = QPushButton("Install", self)
        self.delete_button = QPushButton("Delete", self)
        self.run_button.setObjectName("BtnRun")
        self.install_button.setObjectName("BtnSecondary")
        self.delete_button.setObjectName("BtnDanger")
        button_row.addWidget(self.run_button)
        button_row.addWidget(self.install_button)
        button_row.addWidget(self.delete_button)
        root.addLayout(button_row)

        self.progress = QProgressBar(self)
        self.progress.setRange(0, 100)
        self.progress.setValue(pack.install_progress)
        self.progress.setTextVisible(False)
        self.progress.setFixedHeight(3)
        self.progress_label = QLabel(self)
        self.progress_label.setObjectName("MutedText")
        self.progress_label.setStyleSheet("font-size: 9px;")
        root.addSpacing(9)
        root.addWidget(self.progress)
        root.addSpacing(3)
        root.addWidget(self.progress_label)

        self.apply_status(pack.status)
        self.run_button.clicked.connect(lambda: self.run_requested.emit(self.key))
        self.install_button.clicked.connect(lambda: self.install_requested.emit(self.key))
        self.delete_button.clicked.connect(lambda: self.delete_requested.emit(self.key))

    def _make_tag(self, text: str) -> QLabel:
        tag = QLabel(text, self.tag_row)
        tag.setObjectName("TagPill")
        return tag

    def apply_status(self, status: PackStatus) -> None:
        installing = status == PackStatus.INSTALLING
        self.progress.setVisible(installing)
        self.progress_label.setVisible(installing)
        self.run_button.setEnabled(status in (PackStatus.READY, PackStatus.NEEDS_UPDATE))
        if status == PackStatus.READY:
            self.status_label.setObjectName("StatusReady")
            self.install_button.setText("Install")
        elif status == PackStatus.INSTALLING:
            self.status_label.setObjectName("StatusBusy")
            self.install_button.setText("Reinstall")
        elif status == PackStatus.NEEDS_UPDATE:
            self.status_label.setObjectName("StatusWarn")
            self.install_button.setText("Update")
        else:
            self.status_label.setObjectName("StatusWarn")
            self.install_button.setText("Install")
        self.status_label.style().unpolish(self.status_label)
        self.status_label.style().polish(self.status_label)

    def update_pack(self, pack: Pack) -> None:
        self.name_label.setText(pack.name)
        self.ident_label.setText(pack.abbr())
        self.status_label.setText(status_label(pack.status))
        self.progress.setValue(pack.install_progress)
        self.progress_label.setText(f"Downloading... {pack.install_progress}%")
        self.apply_status(pack.status)
        self.update()

    def set_selected(self, selected: bool) -> None:
        self.setProperty("selected", selected)
        self.style().unpolish(self)
        self.style().polish(self)
        self.update()

    def mousePressEvent(self, event: QMouseEvent) -> None:
        super().mousePressEvent(event)
        self.card_clicked.emit(self.key)

    def paintEvent(self, event) -> None:
        super().paintEvent(event)
        painter = QPainter(self)
        painter.setRenderHint(QPainter.Antialiasing)
        shine = QLinearGradient(0, 0, self.width(), 0)
        shine.setColorAt(0.0, Qt.transparent)
        shine.setColorAt(0.5, QColor(56, 217, 245, 30))
        shine.setColorAt(1.0, Qt.transparent)
        painter.fillRect(0, 0, self.width(), 1, shine)
        glow = QLinearGradient(0, 0, 0, 40)
        glow.setColorAt(0.0, QColor(56, 217, 245, 10))
        glow.setColorAt(1.0, Qt.transparent)
        painter.fillRect(0, 0, self.width(), 40, glow)
